src/execution_tree/primitives/sum_operation.cpp

```diff
@@ -0,0 +1,290 @@
+    {
+        auto m = arg.matrix();
+        blaze::DynamicVector<double> result(m.columns());
+        for (std::size_t i = 0; i < m.columns(); ++i)
+        {
+            auto col = blaze::column(m, i);
+            result[i] = std::accumulate(col.begin(), col.end(), 0.);
+        }
+
+        return primitive_argument_type{result};
+    }
+
+    primitive_argument_type sum_operation::sum2d_axis1(arg_type&& arg) const
+    {
+        auto m = arg.matrix();
+        blaze::DynamicVector<double> result(m.rows());
+        for (std::size_t i = 0; i < m.rows(); ++i)
+        {
+            auto col = blaze::row(m, i);
+            result[i] = std::accumulate(col.begin(), col.end(), 0.);
+        }
+
+        return primitive_argument_type{result};
+    }
+
+    ///////////////////////////////////////////////////////////////////////////
+    hpx::future<primitive_argument_type> sum_operation::eval(
+        std::vector<primitive_argument_type> const& operands,
+        std::vector<primitive_argument_type> const& args) const
+    {
+        if (operands.empty() || operands.size() > 3)
+        {
+            HPX_THROW_EXCEPTION(hpx::bad_parameter,
+                "sum_operation::eval",
+                execution_tree::generate_error_message(
+                    "the sum_operation primitive requires exactly one, two, or "
+                    "three operands",
+                    name_, codename_));
+        }
+
+        for (auto const& i : operands)
+        {
+            if (!valid(i))
+            {
+                HPX_THROW_EXCEPTION(hpx::bad_parameter,
+                    "sum_operation::eval",
+                    execution_tree::generate_error_message(
+                        "the sum_operation primitive requires that the "
+                        "arguments given by the operands array are "
+                        "valid",
+                        name_, codename_));
+            }
+        }
+
+        auto this_ = this->shared_from_this();
+        return hpx::dataflow(
+            hpx::util::unwrapping(
+                [this_](std::vector<primitive_argument_type>&& args)
+                    -> primitive_argument_type {
+                    // Extract axis and keep_dims
+                    // Presence of axis changes behavior for >2d cases
+                    hpx::util::optional<std::int64_t> axis;
+                    bool keep_dims = false;
+
+                    if (args.empty() || args.size() > 3)
+                    {
+                        HPX_THROW_EXCEPTION(hpx::bad_parameter,
+                            "sum_operation::eval",
+                            execution_tree::generate_error_message(
+                                "invalid number of arguments specified", this_->name_,
+                                this_->codename_));
+                    }
+                    else if (args.size() > 1)
+                    {
+                        switch (args[1].index())
```

Comment:
`args[1]` refers to a `primitive_argument_type`. That means that `index == 0, 1, 2` mean `nil`, `node_data<bool>`, and `std::int64_t`. Is that what you need?

Reply:
Yes:
* Argument `axis` may or may not be present
    * case 0 if not present
    * case 2 if present).
* Argument `keep_dims` is also optional, but has to be the 3rd argument if `axis` is also present.
   * case 1
   * nested if inside case 2.

Reply:
I think you should utilize one of the existing extraction functions for this instead (for instance `phylanx::execution_tree::numeric_operand` or similar).

Generally I'd like to avoid having checks for the index() of any `primitive_argument_type` all over the code, Let's try to keep that localized.

Reply:
Done.